### PR TITLE
Documentation: Add graceful restart section in BGP documentation

### PR DESCRIPTION
**Pull request by harsimran-pabla**
Introduce BGP Graceful Restart documentation in bgp-control-plane.rst document.

### CODE REVIEW

File: Documentation/network/bgp-control-plane.rst [hunk 1]
```diff
@@ -281,7 +281,56 @@ values, graceful restart configuration and others.
 
    Change of an existing neighbor configuration can cause reset of the existing BGP
    peering connection, which results in route flaps and transient packet loss while
-   the session reestablishes and peers exchange their routes.
+   the session reestablishes and peers exchange their routes. To prevent packet loss,
+   it is recommended to configure BGP graceful restart.
```
Comment (harsimran-pabla) [inline Documentation/network/bgp-control-plane.rst:285]: @YutaroHayakawa @rastislavs Is this recommendation okay to give in the document?

Comment (rastislavs) [reply Documentation/network/bgp-control-plane.rst:285]: yeah, makes sense to me :)

**Review by rastislavs — APPROVED** [review]
LGTM

File: Documentation/network/bgp-control-plane.rst [hunk 2]
```diff
@@ -281,7 +281,56 @@ values, graceful restart configuration and others.
...
+
+Graceful Restart
+''''''''''''''''
+The Cilium BGP control plane can be configured to act as a graceful restart 
+``Restarting Speaker``. By enabling graceful restart, BGP session will restart and
+graceful restart capability will be advertised in the BGP OPEN message.
```
Comment (qmonnet) [inline Documentation/network/bgp-control-plane.rst]:
```suggestion
``Restarting Speaker``. When you enable graceful restart, the BGP session restarts and
the "graceful restart" capability will be advertised in the BGP OPEN message.
```

File: Documentation/network/bgp-control-plane.rst [hunk 3]
```diff
@@ -281,7 +281,56 @@ values, graceful restart configuration and others.
...
+
+In the event of a cilium agent restart, the peering BGP router will not withdraw 
+routes received from the Cilium BGP control plane immediately. As the datapath
+continues to forward traffic during agent restart, there will be no traffic
+disruption.
```
Comment (qmonnet) [inline Documentation/network/bgp-control-plane.rst]:
```suggestion
In the event of a Cilium Agent restart, the peering BGP router does not withdraw 
routes received from the Cilium BGP control plane immediately. The datapath
continues to forward traffic during Agent restart, so there is no traffic
disruption.
```

File: Documentation/network/bgp-control-plane.rst [hunk 4]
```diff
@@ -281,7 +281,56 @@ values, graceful restart configuration and others.
...
+The Cilium BGP control plane can be configured to act as a graceful restart 
+``Restarting Speaker``. By enabling graceful restart, BGP session will restart and
+graceful restart capability will be advertised in the BGP OPEN message.
+
+In the event of a cilium agent restart, the peering BGP router will not withdraw 
+routes received from the Cilium BGP control plane immediately. As the datapath
+continues to forward traffic during agent restart, there will be no traffic
+disruption. 
+
+The graceful restart capability is configured on per neighbor basis like this:
```
Comment (qmonnet) [inline Documentation/network/bgp-control-plane.rst]:
```suggestion
Configure graceful restart on a per-neighbor basis, as follows:
```

File: Documentation/network/bgp-control-plane.rst [hunk 5]
```diff
@@ -281,7 +281,56 @@ values, graceful restart configuration and others.
...
+      # [...]
+      neighbors: # []CiliumBGPNeighbor
+       - peerAddress: 'fc00:f853:ccd:e793::50/128'
+         # [...]
+         gracefulRestart:
+           enabled: true
+           restartTimeSeconds: 120
+
+.. note::
+   When enabled, graceful restart capability is advertised for IPv4 and IPv6 address families.
```
Comment (qmonnet) [inline Documentation/network/bgp-control-plane.rst]:
```suggestion
.. note::

   When enabled, graceful restart capability is advertised for IPv4 and IPv6 address families.
```

File: Documentation/network/bgp-control-plane.rst [hunk 6]
```diff
@@ -281,7 +281,56 @@ values, graceful restart configuration and others.
...
+
+Users also have an option to set ``RestartTime`` parameter. ``RestartTime`` is the time
+advertised to the peer within which Cilium BGP control plane is expected to re-establish
+the BGP session after a restart. On expiration of ``RestartTime``, peer will remove 
+previously advertised routes by Cilium BGP control plane.
```
Comment (qmonnet) [inline Documentation/network/bgp-control-plane.rst]:
```suggestion
Optionally, you can use the ``RestartTime`` parameter. ``RestartTime`` is the time
advertised to the peer within which Cilium BGP control plane is expected to re-establish
the BGP session after a restart. On expiration of ``RestartTime``, the peer removes 
the routes previously advertised by the Cilium BGP control plane.
```

File: Documentation/network/bgp-control-plane.rst [hunk 7]
```diff
@@ -281,7 +281,56 @@ values, graceful restart configuration and others.
...
+
+When the Cilium agent restarts, TCP FIN will be generated as BGP TCP socket is closed. Peer
+on receiving TCP FIN will change BGP state to ``Idle`` and ``RestartTime`` timer will start.
```
Comment (qmonnet) [inline Documentation/network/bgp-control-plane.rst]:
```suggestion
When the Cilium Agent restarts, it closes the BGP TCP socket, causing the emission of a TCP FIN packet. On receiving this TCP FIN, the peer changes its BGP state to ``Idle`` and starts its ``RestartTime`` timer.
```

File: Documentation/network/bgp-control-plane.rst [hunk 8]
```diff
@@ -281,7 +281,56 @@ values, graceful restart configuration and others.
...
+
+The Cilium agent boot up time varies depending on the deployment. Users are advised to 
+configure ``RestartTime`` greater than time taken by the Cilium agent to boot up.
```
Comment (qmonnet) [inline Documentation/network/bgp-control-plane.rst]:
```suggestion
The Cilium agent boot up time varies depending on the deployment. If using ``RestartTime``, you should set it to a duration greater than the time taken by the Cilium Agent to boot up.
```

**Review by qmonnet — APPROVED** [review]
Looks good. I'm trying to apply some of Sarah's recommendations (2nd person, active voice, present tense, ... I surely missed plenty) and I've got some suggestions, but please treat them as such, and feel free to reject any you don't like.

**Comment by harsimran-pabla** [comment]
Thank you @qmonnet, I have addressed all the comments.

**Review by qmonnet — APPROVED** [review]
OK, then I'm still good :)

**Comment by qmonnet** [comment]
Doc change only, and documentation workflows have passed successfully. This is ready to merge.

**Pull request merged**